rfcs/0001-fobnail-architecture-change.md

```diff
@@ -0,0 +1,143 @@
+   provisioned platforms and all keys.
+
+## Local platform provisioning
+
+Platform provisioning doesn't change much, except that the process is initiated
+by platform instead of Fobnail Token. In order to protect against platform
+re-provisioning itself, Fobnail Token must require user consent to provision
+platform, presumably by pressing button for a few seconds. As an additional
+security measure, platform provisioning could be lockable through Fobnail Token
+APIs.
+
+Previously all requests were sent in a specific order, but since Fobnail Token
+will be a server now, and client is free to send requests in any order, the
+server must handle this case. We want to reduce the amount of global state we
+keep, all API should take as parameter everything that is required to complete
+operation. To achieve this we introduce objects IDs, Fobnail Token may cache
+some objects in memory, like EK or AIK. Client may use object IDs to reference
+these objects. Each object is temporary and is kept at most as long as client is
+connected to the server. Server regularly issues CoAP ping requests to check
+client connectivity.
+
+1. Platform starts provisioning by sending it's EK certificate chain.
+2. Fobnail Token verifies certificate chain and sends EK object ID in response.
+3. Platform generates and sends AIK together with EK object ID.
+4. Fobnail Token responds with AIK challenge and AIK object ID.
+5. Platform performs Credential Activation using the provided challenge and
+   sends results to Fobnail Token together with EK object ID and AIK object ID.
+6. Fobnail verifies results of Credential Activation and sends status to
+   the Platform. AIK is marked as trusted at this point. Fobnail Token creates
+   Provisioning Context and sends its ID (PC ID) to Platform.
+7. Platform signs and sends its metadata to Fobnail, together with PC ID,
+   Fobnail Token verifies metadata signature against AIK bound to specified
+   Provisioning Context.
+8. Fobnail Token verifies metadata and returns status. Metadata is bound to the
+   specified Provisioning Context.
+9. Platform sends RIMs.
+10. Fobnail verifies RIMs and returns status. RIMs are bound to the Provisioning
+    Context.
+11. At this point additional calls may be issued to send more data. This feature
+    is left for future provisioning extensions.
+12. Platform finalizes provisioning by sending special request.
+13. Fobnail verifies all data bound to the Provisioning Context (AIK and RIMs
+    currently) saves it to flash and sends final status.
+14. Platform is provisioned now.
+
+## Remote platform provisioning
+
+During remote platform provisioning Platform Owner is responsible for
+communicating with platform's TPM, verifying TPM (EK certification and AIK
+challenge), receiving and verifying RIMs. Fobnail's sole responsibility is to
+accept configuration provided by Platform Owner which completes provisioning
+process.
```

Comment:
We should probably discuss that. For remote provisioning we need to establish secure communication with PO. In previous architecture messages from PO to Fobnail were signed by PO priv key. However, which key is that? We don't use only a single PO certificate but a chain, shouldn't we allow to sign messages with any key in chain (respecting X509v3 restrictions)?. PO chain would have to be sent every time we provision something.

In original [architecture](https://fobnail.3mdeb.com/images/Fobnail-flows-remote-platform-provisioning.png) we send PO chain to Platform which does nothing with it. Shouldn't PO be sent to Fobnail Token instead?

Reply:
There are a few problems with remote provisioning
- after architecture change, it is platform who requests provisioning, how can PO know whether client is allowed to provision itself? This actually affects old architecture too - communication could be hijacked and PO even wouldn't know that.
- Platform metadata and RIMs are transferred signed but without nonce, which allows replay attacks, still host has to pass Credential Activation, but attack could be done in the middle of communication. This may result in re-provisioning an old, once trusted software, now with known vulnerabilities.
- @Asiderr Where is Fobnail Token located during remote provisioning? Is it plugged into platform being provisioned or into Platform Owner?
- If Fobnail Token is plugged into provisioned machine then we also vulnerable to replay - metadata, RIM and policy sent to token don't have nonce.
- Let alone lack of encryption between remote hosts (PO and Platform), we are risking data leaks.

Reply:
* Initial provisioning should be in `clean room`, Fobnail Token knows it is initial since it has no previous signs of this context (unique data about platform). Whoever kicks client to start provisioning is PO.
* Why not add nonce in communication?
* I would assume remote provisioning means platform is remote to Fobnail Token, so token is connected to terminal operated by PO and PO can start from his terminal remotely on platform e.g. by calling client through SSH.
* it is not
* we should use ssh and operation should be in `clean roon` e.g. trustworthy intranet

Reply:
I started working on remote provisioning (document contains unfinished changes). Generally, both local and remote provisioning should look the same from Attester point of view.

Reply:
@arturkow2000 I'm not sure where you want to go with this conversation. I expressed my point of view and see some stuff is addressed in remote poisoning. Maybe @dpsmith and/or @krystian-hebel should say something.

Reply:
@pietrushnic I wasn't sure how the provisioning should look like (whether FT is plugged into PO or provisioned platform). Now, since my questions have been answerred I defined complete procedure. If you are okay with this procedure I think this can be closed